errors/error.go

```diff
@@ -0,0 +1,373 @@
+// Copyright 2022 Mineiros GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package errors implements the Terramate standard error type.
+// It's heavily influenced by Rob Pike `errors` package in the Upspin project:
+// 	https://commandcenter.blogspot.com/2017/12/error-handling-in-upspin.html
+package errors
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/hcl/v2"
+)
+
+// Error is the default Terramate error type.
+// It may contains some fields unset, but not all.
+// See E() for its usage.
+type Error struct {
+	// Kind is the kind of error.
+	Kind Kind
+
+	// Description of the error.
+	Description string
+
+	// FileRange holds the error source.
+	FileRange hcl.Range
+
+	// Stack which originated the error.
+	Stack StackMeta
+
+	// Err represents the underlying error.
+	Err error
+}
+
+type (
+	// Kind defines the kind of error.
+	Kind string
+	// StackMeta has the metadata of the stack which originated the error.
+	// Same interface as stack.Metadata.
+	StackMeta interface {
+		Name() string
+		Desc() string
+		Path() string
+	}
+)
+
+const separator = ": "
+
+// E builds an error value from its arguments.
+// There must be at least one argument or E panics.
+// The type of each argument determines its meaning. If more than one argument
+// of a given type is presented, only the last one is recorded.
+//
+// The types are:
+// 	errors.Kind
+//		The kind of error (eg.: HCLSyntax, TerramateSchema, etc).
+//	hcl.Range
+//		The file range where the error originated.
+//  errors.Stack
+//		The stack which the error originated.
+//	error
+//		The underlying error that triggered this one.
+//	hcl.Diagnostics
+//		The underlying hcl error that triggered this one.
+//		If this error's FileRange is not set, the diagnostic subject range is
```

Comment:
```suggestion
//		If hcl.Range is not set, the diagnostic subject range is
```

Reply:
Since on the context of the errors.E constructor we are usually talking about the types of the variadic args passed, not the field on the struct they will be stored.